### PR TITLE
Add Readme

### PR DESCRIPTION
**Pull request by Itxaka**
Fixes: https://github.com/kairos-io/kairos/issues/918

### CODE REVIEW

**Review by saamalik — APPROVED** [review]
LGTM! Thanks

File: README.md [hunk 1]
```diff
@@ -19,8 +19,235 @@
...
+It starts pretty early in the boot process, just after `systemd-udev-settle.service` and before `dracut-initqueue.service`
+To see the full bootup process from dracut you can check [here](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html)
+
+Just after starting, Immucore mounts `/proc` if it's not mounted in order to read the `/proc/cmdline` and obtains the different stanzas in order to configure itself.
+After checking the cmdline, it knows in which path is being booted, either active/passive/recovery or Netboot/LiveCD/Do nothing.
+
+Based on that it builds a DAG with the steps needed to complete and process through the DAG until its completed. It also builds a `State` object which has all the configs needed to properly mount and configure the system properly.
+Once the DAG has been completed (and with no errors), Immucore its finished, and it's ready for the initramfs init process to do a switch_root and pivot into the final root to boot the system.
+
+When booting from Netboot/LiveCD/Do nothing (`rd.cos.disable` or `rd.immucore.disable` on the cmdline) the DAD It's pretty simple.
```
Comment (jimmykarily) [inline README.md:184]:
```suggestion
When booting from Netboot/LiveCD/Do nothing (`rd.cos.disable` or `rd.immucore.disable` on the cmdline) the DAG It's pretty simple. 
```

File: README.md [hunk 2]
```diff
@@ -19,8 +19,235 @@
...
+
+----
+
+It starts pretty early in the boot process, just after `systemd-udev-settle.service` and before `dracut-initqueue.service`
+To see the full bootup process from dracut you can check [here](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html)
+
+Just after starting, Immucore mounts `/proc` if it's not mounted in order to read the `/proc/cmdline` and obtains the different stanzas in order to configure itself.
+After checking the cmdline, it knows in which path is being booted, either active/passive/recovery or Netboot/LiveCD/Do nothing.
+
+Based on that it builds a DAG with the steps needed to complete and process through the DAG until its completed. It also builds a `State` object which has all the configs needed to properly mount and configure the system properly.
```
Comment (jimmykarily) [inline README.md:181]:
```suggestion
Based on that it builds a DAG with the steps needed to complete and process through the DAG until its completed. It also builds a `State` object which has all the configs needed to mount and configure the system properly.
```

File: README.md [hunk 3]
```diff
@@ -19,8 +19,235 @@
...
+8.
+ <custom-mount> (background: false) (weak: false)
+ <overlay-mount> (background: false) (weak: false)
+9.
+ <mount-bind> (background: false) (weak: false)
+10.
+ <write-fstab> (background: false) (weak: true)
+```
+
+As show in the DAG, the steps are in order and that shows their dependencies, i.e. `mount-root` depends on `discover-state`and that is why it's just below it.
```
Comment (jimmykarily) [inline README.md:217]:
```suggestion
As shown in the DAG, the steps are in order and that shows their dependencies, i.e. `mount-root` depends on `discover-state`and that is why it's just below it.
```

File: README.md [hunk 4]
```diff
@@ -19,8 +19,235 @@
...
+It won't run until the previous step has completed **without errors**
```
Comment (jimmykarily) [inline README.md:218]:
```suggestion
It won't run until the previous step has completed **without errors**.
```

**Review by jimmykarily — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (Itxaka) [reply README.md:184]: I think my way its better. I mean, the Dads are pretty simple, I am at least :P

File: README.md [hunk 5]
```diff
@@ -19,8 +19,240 @@
...
+Immucore is mostly configured via kernel command line parameters or via the `/run/cos/cos-layout.env` environment file.
+
+These are the read write paths the module mounts as part of the overlay
+ephemeral tmpfs: `/etc`, `/root`, `/home`, `/opt`, `/srv`, `/usr/local`
+and `/var`.
+
+
+## Kernel configuration parameters
+
+---
```
Comment (mauromorales) [inline README.md:44]:
```suggestion
```

Comment (mauromorales) [inline README.md:44]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -19,8 +19,240 @@
...
+* `rd.immucore.oemlabel=<label>`: This option sets the label to search for in order
+  to mount the OEM partition. Defaults to COS_OEM
+  Backwards compatible with the old `rd.cos.oemlabel` stanza.
+
+* `rd.immucore.oemtimeout=<seconds>`: By default we assume the existence of a
+  persistent block device labelled `COS_OEM` which is used to keep some
+  configuration data (mostly cloud-init files). The immutable rootfs tries
+  to mount this device at very early stages of the boot even before applying
+  the immutable rootfs configs. It's done this way to enable to configure the
+  immutable rootfs module within the cloud-init files. As the `COS_OEM` device
```
Comment (mauromorales) [inline README.md:78]:
```suggestion
  the immutable rootfs configs. It's done this way to enable the configuration of the
  immutable rootfs module within the cloud-init files. As the `COS_OEM` device
```

File: README.md [hunk 7]
```diff
@@ -19,8 +19,240 @@
...
+  might not be always present the boot process just continues without failing
```
Comment (mauromorales) [inline README.md:79]:
```suggestion
  might not be always present, the boot process just continues without failing
```

File: README.md [hunk 8]
```diff
@@ -19,8 +19,240 @@
...
+  after a certain timeout. This option configures such a timeout. Defaults to
+  5s.
+  Backwards compatible with the old `rd.cos.oemtimeout` stanza.
+
+* `rd.cos.debugrw`/`rd.immucore.debugrw`: This is a boolean option, true if present, false if not.
+  This option sets the root image to be mounted as a writable device. Note this
```
Comment (mauromorales) [inline README.md:85]:
```suggestion
  This option sets the root image to be mounted as a writable device. Note that this
```

File: README.md [hunk 9]
```diff
@@ -19,8 +19,240 @@
...
+
+
+### Configuration with an environment file
+
+---
+
+The immutable rootfs can be configured with the `/run/cos/cos-layout.env`
+environment file. It is important to note that all the immutable root
+configuration is applied in initrd before switching root and after
+`rootfs` cloud-init stage but before `initramfs` stage. So immutable rootfs
```
Comment (mauromorales) [inline README.md:104]:
```suggestion
`rootfs` cloud-init stage but before `initramfs` stage. So the immutable rootfs
```

File: README.md [hunk 10]
```diff
@@ -19,8 +19,240 @@
...
+configuration via cloud-init using the `/run/cos/cos-layout.env` file is
+only effective if called in any of the `rootfs.before`, `rootfs` or
+`rootfs.after` cloud-init stages.
+
+
+In the environment file few options are available:
+
```
Comment (mauromorales) [inline README.md:111]:
```suggestion
```

File: README.md [hunk 11]
```diff
@@ -19,8 +19,240 @@
...
+The immutable rootfs can be configured with the `/run/cos/cos-layout.env`
+environment file. It is important to note that all the immutable root
+configuration is applied in initrd before switching root and after
+`rootfs` cloud-init stage but before `initramfs` stage. So immutable rootfs
+configuration via cloud-init using the `/run/cos/cos-layout.env` file is
+only effective if called in any of the `rootfs.before`, `rootfs` or
+`rootfs.after` cloud-init stages.
+
+
+In the environment file few options are available:
```
Comment (mauromorales) [inline README.md:110]:
```suggestion
In the environment file, a few options are available:
```

File: README.md [hunk 12]
```diff
@@ -19,8 +19,240 @@
...
+
+
+* `VOLUMES=LABEL=<blk_label>:<mountpoint>`: This variable expects a block device
+  and it mountpoint pair space separated list. The default cOS configuration is:
```
Comment (mauromorales) [inline README.md:114]: not totally sure about this one, maybe `a mountpoint`, feel free to adapt or ignore

```suggestion
  and its mountpoint pair space separated list. The default cOS configuration is:
```

File: README.md [hunk 13]
```diff
@@ -19,8 +19,240 @@
...
+
+  `VOLUMES="LABEL=COS_OEM:/oem LABEL=COS_PERSISTENT:/usr/local"`
+
+* `OVERLAY`: It defines the underlying device for the overlayfs as in
+  `rd.cos.overlay=` kernel parameter.
+
+* `MERGE=true`: Sets makes the `VOLUMES` values to be merged with any other
```
Comment (mauromorales) [inline README.md:121]:
```suggestion
* `MERGE=true`: When set, it makes the `VOLUMES` values to be merged with any other
```

File: README.md [hunk 14]
```diff
@@ -19,8 +19,240 @@
...
+  volume that might have been defined in the kernel command line. The merging
+  criteria is simple: any overlapping volume is overwritten all others are
```
Comment (mauromorales) [inline README.md:123]:
```suggestion
  criteria is simple: any overlapping volume is overwritten, all others are
```

File: README.md [hunk 15]
```diff
@@ -19,8 +19,240 @@
...
+
+* `PERSISTENT_STATE_PATHS`: This is a space separated list of paths. These are
+  the paths that will become writable and store its data inside
+  `PERSISTENT_STATE_TARGET`. By default, this variable is empty, which means
+  no persistent state area is created or used.
+
+  **Note**: The specified paths needs either to exist or be located in an area
+  which is writeable ( for example, inside locations specified with `RW_PATHS`).
+  The dracut module will attempt to create non-existant directories,
+  but might fail if the mountpoint where are located is read-only.
```
Comment (mauromorales) [inline README.md:147]:
```suggestion
  but might fail if the mountpoint where they are located is read-only.
```

File: README.md [hunk 16]
```diff
@@ -19,8 +19,240 @@
...
+
+* `PERSISTENT_STATE_BIND="true|false"`: When this variable is set to true
+  the persistent state paths are bind mounted (instead of using overlayfs)
+  after being mirrored with the original content. By default, this variable is
+  set to `false`.
+
+Note that persistent state are is set up once the ephemeral paths and persistent
```
Comment (mauromorales) [inline README.md:154]:
```suggestion
Note that persistent state is set up once the ephemeral paths and persistent
```

File: README.md [hunk 17]
```diff
@@ -19,8 +19,240 @@
...
+        OVERLAY: "tmpfs:25%"
+```
+
+You can also see the default config that we provide in https://github.com/kairos-io/kairos/blob/master/overlay/files/system/oem/11_persistency.yaml
+
+## What is the default workflow of Immucore
+
+----
+
+It starts pretty early in the boot process, just after `systemd-udev-settle.service` and before `dracut-initqueue.service`
```
Comment (mauromorales) [inline README.md:180]:
```suggestion
It starts pretty early in the boot process, just after `systemd-udev-settle.service` and before `dracut-initqueue.service`.
```

File: README.md [hunk 18]
```diff
@@ -19,8 +19,240 @@
...
+To see the full bootup process from dracut you can check [here](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html)
```
Comment (mauromorales) [inline README.md:181]:
```suggestion
To see the full bootup process from dracut you can check [here](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html).
```

File: README.md [hunk 19]
```diff
@@ -19,8 +19,240 @@
...
+
+Just after starting, Immucore mounts `/proc` if it's not mounted in order to read the `/proc/cmdline` and obtains the different stanzas in order to configure itself.
```
Comment (mauromorales) [inline README.md:183]:
```suggestion
Just after starting, Immucore mounts `/proc` if it's not mounted, it does so in order to read the `/proc/cmdline` and obtains the different stanzas in order to configure itself.
```

File: README.md [hunk 20]
```diff
@@ -19,8 +19,240 @@
...
+After checking the cmdline, it knows in which path is being booted, either active/passive/recovery or Netboot/LiveCD/Do nothing.
+
+Based on that it builds a DAG with the steps needed to complete and process through the DAG until its completed. It also builds a `State` object which has all the configs needed to mount and configure the system properly.
+Once the DAG has been completed (and with no errors), Immucore its finished, and it's ready for the initramfs init process to do a switch_root and pivot into the final root to boot the system.
+
+When booting from Netboot/LiveCD/Do nothing (`rd.cos.disable` or `rd.immucore.disable` on the cmdline) the DAG It's pretty simple. 
+It proceeds to create a sentinel file under `/run/cos/` with the boot mode (`live_mode`) so cloud configs can identify that they are booting from live media and ends.
+
+
+When booting from active/passive/recovery the DAG gets a bit more complicated. You can see the default DAG for an active/passive/recovery system by running immucore with `--dry-run`
```
Comment (mauromorales) [inline README.md:193]:
```suggestion
When booting from active/passive/recovery the DAG gets a bit more complicated. You can see the default DAG for an active/passive/recovery system by running Immucore with `--dry-run`.
```

File: README.md [hunk 21]
```diff
@@ -19,8 +19,240 @@
...
+8.
+ <custom-mount> (background: false) (weak: false)
+ <overlay-mount> (background: false) (weak: false)
+9.
+ <mount-bind> (background: false) (weak: false)
+10.
+ <write-fstab> (background: false) (weak: true)
+```
+
+As shown in the DAG, the steps are in order and that shows their dependencies, i.e. `mount-root` depends on `discover-state`and that is why it's just below it.
```
Comment (mauromorales) [inline README.md:222]:
```suggestion
As shown in the DAG, the steps are in order and that shows their dependencies, i.e. `mount-root` depends on `discover-state` and that is why it's just below it.
```

File: README.md [hunk 22]
```diff
@@ -19,8 +19,240 @@
...
+
+### Steps explained
+
+ - `mount-state`: Will mount the `COS_STATE` partition under `/run/initramfs/cos-state`
+ - `mount-tmpfs`: Will mount `/tmp` 
+ - `create-sentinel`: Will create the sentinel file identifying the boot mode (`active_mode`, `passive_mode`, `recovery_mode` or `live_mode`) under `/run/cos/`
+ - `mount-base-overlay`: Will mount the base overlay under `/run/overlay`
+ - `discover-state`: Will find the correct image under `/run/initramfs/cos-state` and mount it as a loop device
+ - `mount-root`: Will mount the `/dev/disk/by-label/$LABEL` device under the sysroot (Usually `/sysroot`). This label is set in grub depending on the selected entry, as part of the cmdline (i.e. `root=LABEL=COS_ACTIVE`) 
+ - `mount-oem`: Will **try** to mount the oem label device under `/sysroot/oem`. This label is set in grub by default (`rd.cos.oemlabel=COS_OEM`) but also on the default `cos-layout.env` file with Kairos. This partition is not mandatory so It's allowed to fail.
```
Comment (mauromorales) [inline README.md:236]: all others are not using dots, feel free to either remove this one or add to the others

```suggestion
 - `mount-oem`: Will **try** to mount the oem label device under `/sysroot/oem`. This label is set in grub by default (`rd.cos.oemlabel=COS_OEM`) but also on the default `cos-layout.env` file with Kairos. This partition is not mandatory so It's allowed to fail
```

File: README.md [hunk 23]
```diff
@@ -19,8 +19,240 @@
...
+ - `rootfs-hook`: Runs the cloud config stage `rootfs`. Notice that this runs very early in the process so things like binds or RW paths are not yet mounted.
```
Comment (mauromorales) [inline README.md:237]:
```suggestion
 - `rootfs-hook`: Runs the cloud config stage `rootfs`. Notice that this runs very early in the process so things like binds or RW paths are not yet mounted
```

File: README.md [hunk 24]
```diff
@@ -19,8 +19,240 @@
...
+ - `load-config`: This parses the `/run/cos/cos-layout.env` file (usually generated by the `rootfs` stage) and loads all the configurations.
```
Comment (mauromorales) [inline README.md:238]:
```suggestion
 - `load-config`: This parses the `/run/cos/cos-layout.env` file (usually generated by the `rootfs` stage) and loads all the configurations
```

File: README.md [hunk 25]
```diff
@@ -19,8 +19,240 @@
...
+ - `overlay-mount`: This mounts the paths set in the config (`RW_PATHS`) under the `/run/overlay` dir, so they are RW.
```
Comment (mauromorales) [inline README.md:239]: maybe it's the other way around hehe sorry, but feel free to choose

```suggestion
 - `overlay-mount`: This mounts the paths set in the config (`RW_PATHS`) under the `/run/overlay` dir, so they are RW
```

File: README.md [hunk 26]
```diff
@@ -19,8 +19,240 @@
...
+ - `custom-mount`: This mounts the paths set in the config (`VOLUMES`) or in cmdline `rd.cos.mount=` in the given path (`LABEL=COS_PERSISTENT:/usr/local`)
+ - `mount-bind`: This mounts the paths set in the config (`PERSISTENT_STATE_PATHS` and `CUSTOM_BIND_MOUNTS`) as bind mounts under the `PERSISTENT_STATE_TARGET` which defaults to `/usr/local/.state`
+ - `write-fstab`: Writes the final fstab with all the mounts into `/sysroot/fstab`
+
+### UKI mode (Experimental)
+
+---
+
+There is currently support to boot in UKI mode without doing a final switch_root into `/sysroot`
+This means that the initramfs is not really an initramfs but the final system and contains all the needed stuff to boot.
```
Comment (mauromorales) [inline README.md:249]:
```suggestion
This means that the initramfs is not really an initramfs but the final system and contains all the needed parts to boot.
```

File: README.md [hunk 27]
```diff
@@ -19,8 +19,240 @@
...
+This mixed with a UKI binary in which we dump everything into the final binary means that you can have a single EFI file with your full system.
```
Comment (mauromorales) [inline README.md:250]:
```suggestion
This, mixed with a UKI binary in which we dump everything into the final binary, means that you can have a single EFI file with your full system.
```

File: README.md [hunk 28]
```diff
@@ -19,8 +19,240 @@
 </p>
 
 
-With Kairos you can build immutable, bootable Kubernetes and OS images for your edge devices as easily as writing a Dockerfile. Optional P2P mesh with distributed ledger automates node bootstrapping and coordination. Updating nodes is as easy as CI/CD: push a new image to your container registry and let secure, risk-free A/B atomic upgrades do the rest.
+## What is immucore?
```
Comment (mauromorales) [inline README.md:22]:
```suggestion
## What is Immucore?
```

File: README.md [hunk 29]
```diff
@@ -19,8 +19,240 @@
...
 
+---
+
+Immucore is the management interface to mount Kairos disks and filesystems.
+Is a dracut module responsible for mounting the root tree during boot time with the immutable specific setup.
```
Comment (oz123) [inline README.md:27]: It is a dracut module...

Comment (oz123) [reply README.md:27]: ... with the specific immutable setup.

File: README.md [hunk 30]
```diff
@@ -19,8 +19,240 @@
...
+The immutability concept refers to read only root (/) system.
+To ensure the linux OS is still functional certain paths or areas are required to be writable,
```
Comment (oz123) [inline README.md:29]: To ensure ... functional, certain file system path must be writable.

File: README.md [hunk 31]
```diff
@@ -19,8 +19,240 @@
...
+in those cases an ephemeral overlay tmpfs is set in place.
```
Comment (oz123) [inline README.md:30]: Please explain what ephemral ...
e.g. (ephemeral - meaning chages are delete on reboot - ...

File: README.md [hunk 32]
```diff
@@ -19,8 +19,240 @@
...
+
+## Kernel configuration parameters
+
+---
+
+The immutable rootfs can be configured with the following kernel parameters:
+
+* `cos-img/filename=<imgfile>`: This is one of the main parameters, it defines
+  the location of the image file to boot from. This defines the booting mode for
+  Immucore, setting in motion the full DAG to set up the system.
```
Comment (oz123) [inline README.md:50]: Jargon - DAG is first mentioned here. The naive user does not know what this means.
Can this be replaced?

File: README.md [hunk 33]
```diff
@@ -19,8 +19,240 @@
...
+* `rd.immucore.overlay=tmpfs:<size>`: This defines the size of the tmpfs used for
+  the ephemeral overlayfs. It can be expressed in MiB or as a % of the available
+  memory. Defaults to `rd.immucore.overlay=tmpfs:20%` if not present.
+  Backwards compatible with the old `rd.cos.overlay` stanza.
+
+* `rd.immucore.overlay=LABEL=<vol_label>`: Optionally and mostly for debugging
+  purposes the overlayfs can be mounted on top of a persistent block device.
+  Block devices can be expressed by LABEL (`LABEL=<blk_label>`) or by UUID
+  (`UUID=<blk_uuid>`)
+  Backwards compatible with the old `rd.cos.overlay` stanza.
```
Comment (oz123) [inline README.md:61]: IMHO, here is `directive` is a better term compared to stanza.

File: README.md (hunk 12 above)
Comment (oz123) [reply README.md:114]: moutpoint -> mount point.

File: README.md [hunk 34]
```diff
@@ -19,8 +19,240 @@
...
+It starts pretty early in the boot process, just after `systemd-udev-settle.service` and before `dracut-initqueue.service`
+To see the full bootup process from dracut you can check [here](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html)
+
+Just after starting, Immucore mounts `/proc` if it's not mounted in order to read the `/proc/cmdline` and obtains the different stanzas in order to configure itself.
+After checking the cmdline, it knows in which path is being booted, either active/passive/recovery or Netboot/LiveCD/Do nothing.
+
+Based on that it builds a DAG with the steps needed to complete and process through the DAG until its completed. It also builds a `State` object which has all the configs needed to mount and configure the system properly.
+Once the DAG has been completed (and with no errors), Immucore its finished, and it's ready for the initramfs init process to do a switch_root and pivot into the final root to boot the system.
+
+When booting from Netboot/LiveCD/Do nothing (`rd.cos.disable` or `rd.immucore.disable` on the cmdline) the DAG It's pretty simple.
```
Comment (oz123) [inline README.md:189]: ... the DAG is pretty simple.

File: README.md [hunk 35]
```diff
@@ -19,8 +19,240 @@
...
+ - `overlay-mount`: This mounts the paths set in the config (`RW_PATHS`) under the `/run/overlay` dir, so they are RW.
+ - `custom-mount`: This mounts the paths set in the config (`VOLUMES`) or in cmdline `rd.cos.mount=` in the given path (`LABEL=COS_PERSISTENT:/usr/local`)
+ - `mount-bind`: This mounts the paths set in the config (`PERSISTENT_STATE_PATHS` and `CUSTOM_BIND_MOUNTS`) as bind mounts under the `PERSISTENT_STATE_TARGET` which defaults to `/usr/local/.state`
+ - `write-fstab`: Writes the final fstab with all the mounts into `/sysroot/fstab`
+
+### UKI mode (Experimental)
+
+---
+
+There is currently support to boot in UKI mode without doing a final switch_root into `/sysroot`
```
Comment (oz123) [inline README.md:248]: Currently, there is ... final `switch_root` into ...

File: README.md (hunk 26 above)
Comment (oz123) [reply README.md:249]: This means that the initramfs is not really an initramfs. Nevertheless, the final system contains all the ...

**Review by oz123 — CHANGES_REQUESTED** [review]

**Review by oz123 — APPROVED** [review]
LGTM.

**Pull request merged**